### ISSUE TITLE
ERROR: 1038 (HY001): Out of sort memory, consider increasing server sort buffer size

### ISSUE THREAD

**Issue by muyesh** [tjni/langgraph-checkpoint-mysql#34]
I would like to report an issue I have encountered. If anyone has any good ideas or advice, I would greatly appreciate it.

MySQL: 8.0.40
langgraph-checkpoint-mysql: 2.0.3

Currently, in the Bot system I developed for use with Microsoft Teams, I am encountering an error like the one mentioned in the title when loading checkpoints. After investigating the cause, it seems that the error is due to the sort buffer being exhausted when using JSON columns. Therefore, I reused the SELECT_SQL in the code and conducted tests using the following three SQL statements.

```SQL
select
  thread_id,
  checkpoint,
  checkpoint_ns,
  checkpoint_id,
  parent_checkpoint_id,
  metadata,
  (
      select json_arrayagg(json_array(bl.channel, bl.type, bl.blob))
      from
      (
          select channel, json_unquote(
              json_extract(checkpoint, concat('$.channel_versions.', '"', channel, '"'))
          ) as version
          from json_table(
              json_keys(checkpoint, '$.channel_versions'),
              '$[*]' columns (channel VARCHAR(150) PATH '$')
          ) as channels
      ) as channel_versions
      inner join checkpoint_blobs bl
          on bl.thread_id = checkpoints.thread_id
          and bl.checkpoint_ns = checkpoints.checkpoint_ns
          and bl.channel = channel_versions.channel
          and bl.version = channel_versions.version
  ) as channel_values
from checkpoints
WHERE thread_id = 11 ORDER BY checkpoint_id DESC LIMIT 1
```

```SQL
select
  thread_id,
  checkpoint,
  checkpoint_ns,
  checkpoint_id,
  parent_checkpoint_id,
  metadata,
  (
      select
      json_arrayagg(json_array(cw.task_id, cw.channel, cw.type, cw.blob, cw.idx))
      from checkpoint_writes cw
      where cw.thread_id = checkpoints.thread_id
          and cw.checkpoint_ns = checkpoints.checkpoint_ns
          and cw.checkpoint_id = checkpoints.checkpoint_id
  ) as pending_writes
from checkpoints
WHERE thread_id = 11 ORDER BY checkpoint_id DESC LIMIT 1
```

```SQL
select
  thread_id,
  checkpoint,
  checkpoint_ns,
  checkpoint_id,
  parent_checkpoint_id,
  metadata,
  (
      select json_arrayagg(json_array(cw.task_id, cw.type, cw.blob, cw.idx))
      from checkpoint_writes cw
      where cw.thread_id = checkpoints.thread_id
          and cw.checkpoint_ns = checkpoints.checkpoint_ns
          and cw.checkpoint_id = checkpoints.parent_checkpoint_id
          and cw.channel = '{TASKS}'
  ) as pending_sends
from checkpoints
WHERE thread_id = 11 ORDER BY checkpoint_id DESC LIMIT 1
```

As a result, it seems that the "Out of sort memory" error occurs in each query due to the subquery portion.
Therefore, when I changed the MySQL sort_buffer_size from the default 256KB to 8MB, all the queries started working correctly.

```
MySQL  localhost:3306 ssl  SQL > show variables like 'sort_buffer_size';
+------------------+--------+
| Variable_name    | Value  |
+------------------+--------+
| sort_buffer_size | 262144 |
+------------------+--------+
1 row in set (0.0014 sec)
 MySQL  localhost:3306 ssl  SQL > SET SESSION sort_buffer_size = 8388608;
Query OK, 0 rows affected (0.0005 sec)
 MySQL  localhost:3306 ssl  SQL > show variables like 'sort_buffer_size';
+------------------+---------+
| Variable_name    | Value   |
+------------------+---------+
| sort_buffer_size | 8388608 |
+------------------+---------+
1 row in set (0.0014 sec)
```

I believe the root cause of this issue is the generation of numerous checkpoints under the same thread_id. This stems from my implementation continually using a single thread to accommodate Teams' limited UI.

As a temporary measure, increasing the sort_buffer_size can address the issue, but it will likely hit the limit again eventually.

The only solution I can think of right now is to copy the state to a new thread to reduce the checkpoint history. Is it possible to manually reduce the checkpoint records, or perhaps address the issue with some kind of magical query adjustment?

Does anyone have any good suggestions?

-- added 

Apologies, the issue was occurring regardless of the subquery.

```SQL
select
  thread_id,
  checkpoint,
  checkpoint_ns,
  checkpoint_id,
  parent_checkpoint_id,
  metadata
from checkpoints
WHERE thread_id = 11 ORDER BY checkpoint_id DESC LIMIT 1
```

**Comment by muyesh** [comment]
```
 MySQL  localhost:3306 ssl  aibot_checkpoint  SQL > explain select   thread_id,   checkpoint,   checkpoint_ns,   checkpoint_id,   parent_checkpoint_id,   metadata from checkpoints WHERE thread_id = 11 ORDER BY checkpoint_id DESC LIMIT 1;
+----+-------------+-------------+------------+------+---------------+------+---------+------+------+----------+-----------------------------+
| id | select_type | table       | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra                       |
+----+-------------+-------------+------------+------+---------------+------+---------+------+------+----------+-----------------------------+
|  1 | SIMPLE      | checkpoints | NULL       | ALL  | PRIMARY       | NULL | NULL    | NULL | 6573 |       10 | Using where; Using filesort |
+----+-------------+-------------+------------+------+---------------+------+---------+------+------+----------+-----------------------------+
1 row in set, 3 warnings (0.0007 sec)
Warning (code 1739): Cannot use ref access on index 'PRIMARY' due to type or collation conversion on field 'thread_id'
Warning (code 1739): Cannot use range access on index 'PRIMARY' due to type or collation conversion on field 'thread_id'
Note (code 1003): /* select#1 */ select `aibot_checkpoint`.`checkpoints`.`thread_id` AS `thread_id`,`aibot_checkpoint`.`checkpoints`.`checkpoint` AS `checkpoint`,`aibot_checkpoint`.`checkpoints`.`checkpoint_ns` AS `checkpoint_ns`,`aibot_checkpoint`.`checkpoints`.`checkpoint_id` AS `checkpoint_id`,`aibot_checkpoint`.`checkpoints`.`parent_checkpoint_id` AS `parent_checkpoint_id`,`aibot_checkpoint`.`checkpoints`.`metadata` AS `metadata` from `aibot_checkpoint`.`checkpoints` where (`aibot_checkpoint`.`checkpoints`.`thread_id` = 11) order by `aibot_checkpoint`.`checkpoints`.`checkpoint_id` desc limit 1
```

Therefore, when I added an index to the checkpoint_id, everything started working well.

```SQL
CREATE INDEX idx_checkpoint_id ON checkpoints (checkpoint_id);
```

Here is the EXPLAIN output after adding the index.

```
 MySQL  localhost:3306 ssl  aibot_checkpoint  SQL > explain select   thread_id,   checkpoint,   checkpoint_ns,   checkpoint_id,   parent_checkpoint_id,   metadata from checkpoints WHERE thread_id = 11 ORDER BY checkpoint_id DESC LIMIT 1;         
+----+-------------+-------------+------------+-------+---------------+-------------------+---------+------+------+----------+----------------------------------+                                                                                     
| id | select_type | table       | partitions | type  | possible_keys | key               | key_len | ref  | rows | filtered | Extra                            |                                                                                     
+----+-------------+-------------+------------+-------+---------------+-------------------+---------+------+------+----------+----------------------------------+                                                                                     
|  1 | SIMPLE      | checkpoints | NULL       | index | PRIMARY       | idx_checkpoint_id | 602     | NULL |    1 |       10 | Using where; Backward index scan |                                                                                     
+----+-------------+-------------+------------+-------+---------------+-------------------+---------+------+------+----------+----------------------------------+                                                                                     
1 row in set, 4 warnings (0.0007 sec)                                                                                                                                                                                                                 
Warning (code 1739): Cannot use ref access on index 'PRIMARY' due to type or collation conversion on field 'thread_id'                                                                                                                                
Warning (code 1739): Cannot use range access on index 'PRIMARY' due to type or collation conversion on field 'thread_id'                                                                                                                              
Warning (code 1739): Cannot use range access on index 'idx_checkpoint_id' due to type or collation conversion on field 'thread_id'                                                                                                                    
Note (code 1003): /* select#1 */ select `aibot_checkpoint`.`checkpoints`.`thread_id` AS `thread_id`,`aibot_checkpoint`.`checkpoints`.`checkpoint` AS `checkpoint`,`aibot_checkpoint`.`checkpoints`.`checkpoint_ns` AS `checkpoint_ns`,`aibot_checkpoin
t`.`checkpoints`.`checkpoint_id` AS `checkpoint_id`,`aibot_checkpoint`.`checkpoints`.`parent_checkpoint_id` AS `parent_checkpoint_id`,`aibot_checkpoint`.`checkpoints`.`metadata` AS `metadata` from `aibot_checkpoint`.`checkpoints` where (`aibot_ch
eckpoint`.`checkpoints`.`thread_id` = 11) order by `aibot_checkpoint`.`checkpoints`.`checkpoint_id` desc limit 1
```

How about adding an index to the checkpoint_id using a CREATE TABLE statement?

```SQL
CREATE TABLE checkpoints (
    thread_id VARCHAR(150) NOT NULL,
    checkpoint_ns VARCHAR(150) NOT NULL,
    checkpoint_id VARCHAR(150) NOT NULL,
    parent_checkpoint_id VARCHAR(150),
    type VARCHAR(150),
    checkpoint JSON NOT NULL,
    metadata JSON NOT NULL DEFAULT '{}',
    PRIMARY KEY (thread_id, checkpoint_ns, checkpoint_id),
    INDEX idx_checkpoint_id (checkpoint_id)
);
```

**Comment by tjni** [comment]
Released 2.0.7 with your fix. Thank you so much!
